variables/variables.go

```diff
@@ -0,0 +1,292 @@
+package variables
+
+import (
+	"fmt"
+	"github.com/gruntwork-io/boilerplate/util"
+)
+
+// An interface for a variable defined in a boilerplate.yml config file
+type Variable interface {
```

Comment:
The rest of the time in this PR was spent on this thing. It used to be a struct, but that led to lots of bugs where I was forgetting to set certain values in the struct (mainly, the new `Type` field). Since there is no way to enforce a default value in a struct, I converted it to an interface. 

In a better programming language, this would've been a perfect use case for multiple implementations of this interface: one for string variables, one for int variables, one for enums, and so on. Each implementation would only use the types relevant for it and would never return null or any of that nonsense for fields that didn't apply (e.g. only Enum has an `Options` field). In other words, you use the type system to make illegal types unrepresentable.

In go, this is somewhat possible, but it took a MASSIVE amount of code due to the lack of support for generics, self types, and other useful features you get in better type systems. The result hideous and definitely reduced the maintainability of the project, so I switched back to a single interface and a single implementation. It's not as good as it should be, but hopefully good enough for now.